File: Tests/DraftRightTests/RewriteCacheTests.swift
import XCTest
@testable import DraftRight

/// Client-side rewrite cache (#147). Cases mirror
/// DraftRightWindows.PureTests/RewriteCacheTests.cs so the three
/// implementations cannot drift.
final class RewriteCacheTests: XCTestCase {

    private let tone = "polished"

    // Each test builds its own instance rather than touching `.shared`, so
    // there is no cross-test state and eviction can be exercised without
    // inserting 200 entries.
    private func makeCache(max: Int = RewriteCache.defaultMaxEntries,
                           evict: Int = RewriteCache.defaultEvictFraction) -> RewriteCache {
        RewriteCache(maxEntries: max, evictFraction: evict)
    }

    func testGetReturnsNilOnMiss() {
        XCTAssertNil(makeCache().get(text: "hello", tone: tone))
    }

    func testSetThenGetRoundTrips() {
        let c = makeCache()
        c.set(text: "hello", tone: tone, result: "Hello.")
        XCTAssertEqual(c.get(text: "hello", tone: tone), "Hello.")
    }

    func testDifferentToneIsADifferentEntry() {
        let c = makeCache()
        c.set(text: "hello", tone: "polished", result: "Polished.")
        c.set(text: "hello", tone: "concise", result: "Concise.")
        XCTAssertEqual(c.get(text: "hello", tone: "polished"), "Polished.")
        XCTAssertEqual(c.get(text: "hello", tone: "concise"), "Concise.")
    }

    /// The bug this guards (#147): the old key omitted the language entirely,
    /// so translating to one language and then switching served the previous
    /// language's result — wrong output, not merely a missed hit.
    func testLanguageIsPartOfTheKey() {
        let c = makeCache()
        c.set(text: "hello", tone: "translate", result: "Xin chào", language: "Vietnamese")

        XCTAssertEqual(c.get(text: "hello", tone: "translate", language: "Vietnamese"), "Xin chào")
        XCTAssertNil(c.get(text: "hello", tone: "translate", language: "Japanese"))
        XCTAssertNil(c.get(text: "hello", tone: "translate"))
    }

    func testReSettingSameKeyOverwritesWithoutGrowing() {
        let c = makeCache()
        c.set(text: "hello", tone: tone, result: "first")
        c.set(text: "hello", tone: tone, result: "second")
        XCTAssertEqual(c.get(text: "hello", tone: tone), "second")
        XCTAssertEqual(c.count, 1)
    }

    func testClearDropsEverything() {
        let c = makeCache()
        c.set(text: "a", tone: tone, result: "A")
        c.set(text: "b", tone: tone, result: "B")
        c.clear()
        XCTAssertEqual(c.count, 0)
        XCTAssertNil(c.get(text: "a", tone: tone))
    }

    /// The previous implementation evicted `cache.keys.prefix(...)`. Swift
    /// dictionaries have no key order, so it dropped an arbitrary quarter
    /// while claiming to drop the oldest. This pins real FIFO.
    func testEvictsOldestSliceWhenFull() {
        let c = makeCache(max: 8, evict: 4)   // drop the 2 oldest on the 9th
        for i in 0..<8 { c.set(text: "t\(i)", tone: tone, result: "r\(i)") }
        XCTAssertEqual(c.count, 8)

        c.set(text: "t8", tone: tone, result: "r8")

        XCTAssertEqual(c.count, 7)                                  // 8 - 2 + 1
        XCTAssertNil(c.get(text: "t0", tone: tone))                 // oldest two gone
        XCTAssertNil(c.get(text: "t1", tone: tone))
        XCTAssertEqual(c.get(text: "t2", tone: tone), "r2")         // third-oldest survives
        XCTAssertEqual(c.get(text: "t8", tone: tone), "r8")         // newcomer present
    }

    /// FIFO by insertion, not by last use — matching Windows and Linux.
    /// Documented so a future switch to LRU is deliberate, not accidental.
    func testEvictionOrderIsInsertionNotLastUse() {
        let c = makeCache(max: 4, evict: 4)
        for i in 0..<4 { c.set(text: "t\(i)", tone: tone, result: "r\(i)") }
        _ = c.get(text: "t0", tone: tone)          // touch the oldest
        c.set(text: "t4", tone: tone, result: "r4")

        XCTAssertNil(c.get(text: "t0", tone: tone))  // still evicted despite the read
    }

    func testReSettingExistingKeyDoesNotChangeItsEvictionPosition() {
        let c = makeCache(max: 4, evict: 4)
        for i in 0..<4 { c.set(text: "t\(i)", tone: tone, result: "r\(i)") }
        c.set(text: "t0", tone: tone, result: "updated")  // overwrite oldest
        c.set(text: "t4", tone: tone, result: "r4")       // evicts 1

        XCTAssertNil(c.get(text: "t0", tone: tone))       // t0 was still oldest
        XCTAssertEqual(c.get(text: "t1", tone: tone), "r1")
    }

    func testBoundsAreClampedToAtLeastOne() {
        for bad in [0, -5] {
            let c = makeCache(max: bad)
            c.set(text: "a", tone: tone, result: "A")
            XCTAssertEqual(c.get(text: "a", tone: tone), "A")
            c.set(text: "b", tone: tone, result: "B")
            XCTAssertEqual(c.get(text: "b", tone: tone), "B")
        }
        let c = makeCache(max: 4, evict: 0)
        for i in 0..<5 { c.set(text: "t\(i)", tone: tone, result: "r\(i)") }
        XCTAssertEqual(c.get(text: "t4", tone: tone), "r4")
    }

    func testDefaultsMatchTheOtherClients() {
        XCTAssertEqual(RewriteCache.defaultMaxEntries, 200)
        XCTAssertEqual(RewriteCache.defaultEvictFraction, 4)
    }

    func testKeyShapeMatchesWindowsAndLinux() {
        XCTAssertEqual(RewriteCache.key(text: "hi", tone: "polished"), "polished::hi")
        XCTAssertEqual(RewriteCache.key(text: "hi", tone: "translate", language: "Spanish"),
                       "Spanish::translate::hi")
        // Empty language is treated as absent, not as a distinct bucket.
        XCTAssertEqual(RewriteCache.key(text: "hi", tone: "polished", language: ""), "polished::hi")
    }

    // MARK: - Tone gating

    func testOnlyTranslateUsesTargetLanguage() {
        XCTAssertTrue(Tone.translate.usesTargetLanguage)
        for t in Tone.allCases where t != .translate {
            XCTAssertFalse(t.usesTargetLanguage, "\(t) should not use a target language")
        }
    }

    func testOnlyGrammarCheckProducesUncacheableOutput() {
        XCTAssertFalse(Tone.grammarCheck.producesCacheableText)
        for t in Tone.allCases where t != .grammarCheck {
            XCTAssertTrue(t.producesCacheableText, "\(t) should be cacheable")
        }
    }
}

File: DraftRightWindows/DraftRightWindows.PureTests/RewriteCacheTests.cs
using DraftRightWindows.Services;
using Xunit;

namespace DraftRightWindows.PureTests;

/// <summary>
/// Client-side rewrite cache (issue #108). Mirrors the macOS and Linux
/// implementations: key = (language, tone, text), bounded, FIFO eviction.
/// </summary>
public class RewriteCacheTests
{
    private const string Tone = "polished";

    [Fact]
    public void Get_ReturnsNull_OnMiss()
    {
        var c = new RewriteCache();
        Assert.Null(c.Get("hello", Tone));
    }

    [Fact]
    public void Set_ThenGet_RoundTrips()
    {
        var c = new RewriteCache();
        c.Set("hello", Tone, "Hello.");
        Assert.Equal("Hello.", c.Get("hello", Tone));
    }

    [Fact]
    public void DifferentTone_IsADifferentEntry()
    {
        var c = new RewriteCache();
        c.Set("hello", "polished", "Polished.");
        c.Set("hello", "concise", "Concise.");
        Assert.Equal("Polished.", c.Get("hello", "polished"));
        Assert.Equal("Concise.", c.Get("hello", "concise"));
    }

    [Fact]
    public void Language_IsPartOfTheKey()
    {
        // The bug this guards: storing without the language but reading with it
        // (or vice versa) means translations never hit — and worse, switching
        // language could serve the previous language's result.
        var c = new RewriteCache();
        c.Set("hello", "translate", "Hola.", "Spanish");

        Assert.Equal("Hola.", c.Get("hello", "translate", "Spanish"));
        Assert.Null(c.Get("hello", "translate", "French"));
        Assert.Null(c.Get("hello", "translate"));
    }

    [Fact]
    public void ReSettingSameKey_OverwritesWithoutGrowing()
    {
        var c = new RewriteCache();
        c.Set("hello", Tone, "first");
        c.Set("hello", Tone, "second");
        Assert.Equal("second", c.Get("hello", Tone));
        Assert.Equal(1, c.Count);
    }

    [Fact]
    public void Clear_DropsEverything()
    {
        var c = new RewriteCache();
        c.Set("a", Tone, "A");
        c.Set("b", Tone, "B");
        c.Clear();
        Assert.Equal(0, c.Count);
        Assert.Null(c.Get("a", Tone));
    }

    [Fact]
    public void EvictsOldestSlice_WhenFull()
    {
        // max 8, evict fraction 4 → drop the 2 oldest when the 9th arrives.
        var c = new RewriteCache(maxEntries: 8, evictFraction: 4);
        for (var i = 0; i < 8; i++) c.Set($"t{i}", Tone, $"r{i}");
        Assert.Equal(8, c.Count);

        c.Set("t8", Tone, "r8");

        Assert.Equal(7, c.Count);              // 8 - 2 evicted + 1 inserted
        Assert.Null(c.Get("t0", Tone));        // oldest two gone
        Assert.Null(c.Get("t1", Tone));
        Assert.Equal("r2", c.Get("t2", Tone)); // third-oldest survives
        Assert.Equal("r8", c.Get("t8", Tone)); // newcomer present
    }

    [Fact]
    public void EvictionOrder_IsInsertion_NotLastUse()
    {
        // FIFO, matching macOS and Linux — re-reading an entry does not make it
        // survive eviction. Documented so a future switch to LRU is a conscious
        // change rather than an accident.
        var c = new RewriteCache(maxEntries: 4, evictFraction: 4);
        for (var i = 0; i < 4; i++) c.Set($"t{i}", Tone, $"r{i}");
        c.Get("t0", Tone);          // touch the oldest
        c.Set("t4", Tone, "r4");    // triggers eviction of 1 entry

        Assert.Null(c.Get("t0", Tone));   // still evicted despite the read
    }

    [Fact]
    public void ReSettingExistingKey_DoesNotChangeItsEvictionPosition()
    {
        var c = new RewriteCache(maxEntries: 4, evictFraction: 4);
        for (var i = 0; i < 4; i++) c.Set($"t{i}", Tone, $"r{i}");
        c.Set("t0", Tone, "updated");   // overwrite oldest — must not requeue it
        c.Set("t4", Tone, "r4");        // evicts 1

        Assert.Null(c.Get("t0", Tone));         // t0 was still the oldest
        Assert.Equal("r1", c.Get("t1", Tone));
    }

    [Theory]
    [InlineData(0)]
    [InlineData(-5)]
    public void MaxEntries_IsClampedToAtLeastOne(int badMax)
    {
        // A zero/negative bound must not make Set throw or loop forever.
        var c = new RewriteCache(maxEntries: badMax);
        c.Set("a", Tone, "A");
        Assert.Equal("A", c.Get("a", Tone));
        c.Set("b", Tone, "B");
        Assert.Equal("B", c.Get("b", Tone));
    }

    [Fact]
    public void EvictFraction_IsClampedToAtLeastOne()
    {
        var c = new RewriteCache(maxEntries: 4, evictFraction: 0);
        for (var i = 0; i < 5; i++) c.Set($"t{i}", Tone, $"r{i}");
        Assert.Equal("r4", c.Get("t4", Tone));
    }

    [Fact]
    public void DefaultsMatchTheOtherClients()
    {
        // macOS: maxEntries = 200, evicts maxEntries/4.
        // Linux: REWRITE_CACHE_MAX_ENTRIES = 200, EVICT_FRACTION = 4.
        Assert.Equal(200, RewriteCache.DefaultMaxEntries);
        Assert.Equal(4, RewriteCache.DefaultEvictFraction);
    }

    [Fact]
    public void KeyShape_MatchesMacOsAndLinux()
    {
        Assert.Equal("polished::hello", RewriteCache.Key("hello", "polished"));
        Assert.Equal("Spanish::translate::hello", RewriteCache.Key("hello", "translate", "Spanish"));
        // Empty language is treated as absent, not as a distinct bucket.
        Assert.Equal("polished::hello", RewriteCache.Key("hello", "polished", ""));
    }
}
